File: Design-Patterns-Using-Go/Factory/Factory.go
package main

import "fmt"

// Animal interface
type Animal interface {
	Speak()
}

// Concrete Product classes
type Dog struct{}

func (d *Dog) Speak() {
	fmt.Println("Woof!")
}

type Cat struct{}

func (c *Cat) Speak() {
	fmt.Println("Meow!")
}

// Creator abstract class
type AnimalFactory interface {
	CreateAnimal() Animal
}

// Concrete Creator classes
type DogFactory struct{}

func (df *DogFactory) CreateAnimal() Animal {
	return &Dog{}
}

type CatFactory struct{}

func (cf *CatFactory) CreateAnimal() Animal {
	return &Cat{}
}

// Client code
func main() {
	dogFactory := &DogFactory{}
	dog := dogFactory.CreateAnimal()
	dog.Speak()

	catFactory := &CatFactory{}
	cat := catFactory.CreateAnimal()
	cat.Speak()
}

/*
Woof!
Meow!
*/

File: Design-Patterns-Using-JavaScript/Factory/Factory.js
// Product interface
class Animal {
    speak() {
        throw new Error("Abstract method 'speak' must be implemented.");
    }
}

// Concrete Product classes
class Dog extends Animal {
    speak() {
        console.log("Woof!");
    }
}

class Cat extends Animal {
    speak() {
        console.log("Meow!");
    }
}

// Creator abstract class
class AnimalFactory {
    createAnimal() {
        throw new Error("Abstract method 'createAnimal' must be implemented.");
    }
}

// Concrete Creator classes
class DogFactory extends AnimalFactory {
    createAnimal() {
        return new Dog();
    }
}

class CatFactory extends AnimalFactory {
    createAnimal() {
        return new Cat();
    }
}

// Client code
const dogFactory = new DogFactory();
const dog = dogFactory.createAnimal();
dog.speak();

const catFactory = new CatFactory();
const cat = catFactory.createAnimal();
cat.speak();

/*
Woof!
Meow!
*/
